### ISSUE TITLE
DQN with softmax in the final layer

### ISSUE THREAD

**Issue by link-zju** [aai-institute/tianshou#725]
- [x] I have marked all applicable categories:
    + [x] exception-raising bug
    + [x] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
 ```python
    import argparse
    import os
    import pprint
    
    import gym
    import numpy as np
    import torch
    from torch.utils.tensorboard import SummaryWriter
    
    from tianshou.data import Collector, PrioritizedVectorReplayBuffer, VectorReplayBuffer
    from tianshou.env import DummyVectorEnv
    from tianshou.policy import DQNPolicy
    from tianshou.trainer import offpolicy_trainer
    from tianshou.utils import TensorboardLogger
    from tianshou.utils.net.common import Net
    from tianshou.utils.net.discrete import Actor
    
    
    def get_args():
        parser = argparse.ArgumentParser()
        parser.add_argument('--task', type=str, default='CartPole-v0')
        parser.add_argument('--reward-threshold', type=float, default=None)
        parser.add_argument('--seed', type=int, default=1626)
        parser.add_argument('--eps-test', type=float, default=0.05)
        parser.add_argument('--eps-train', type=float, default=0.1)
        parser.add_argument('--buffer-size', type=int, default=20000)
        parser.add_argument('--lr', type=float, default=1e-3)
        parser.add_argument('--gamma', type=float, default=0.9)
        parser.add_argument('--n-step', type=int, default=3)
        parser.add_argument('--target-update-freq', type=int, default=320)
        parser.add_argument('--epoch', type=int, default=20)
        parser.add_argument('--step-per-epoch', type=int, default=10000)
        parser.add_argument('--step-per-collect', type=int, default=10)
        parser.add_argument('--update-per-step', type=float, default=0.1)
        parser.add_argument('--batch-size', type=int, default=64)
        parser.add_argument(
            '--hidden-sizes', type=int, nargs='*', default=[128, 128, 128, 128]
        )
        parser.add_argument('--training-num', type=int, default=10)
        parser.add_argument('--test-num', type=int, default=100)
        parser.add_argument('--logdir', type=str, default='log')
        parser.add_argument('--render', type=float, default=0.)
        parser.add_argument('--prioritized-replay', action="store_true", default=False)
        parser.add_argument('--alpha', type=float, default=0.6)
        parser.add_argument('--beta', type=float, default=0.4)
        parser.add_argument(
            '--device', type=str, default='cuda' if torch.cuda.is_available() else 'cpu'
        )
        args = parser.parse_known_args()[0]
        return args
    
    
    def test_dqn(args=get_args()):
        env = gym.make(args.task)
        args.state_shape = env.observation_space.shape or env.observation_space.n
        args.action_shape = env.action_space.shape or env.action_space.n
        if args.reward_threshold is None:
            default_reward_threshold = {"CartPole-v0": 195}
            args.reward_threshold = default_reward_threshold.get(
                args.task, env.spec.reward_threshold
            )
        # train_envs = gym.make(args.task)
        # you can also use tianshou.env.SubprocVectorEnv
        train_envs = DummyVectorEnv(
            [lambda: gym.make(args.task) for _ in range(args.training_num)]
        )
        # test_envs = gym.make(args.task)
        test_envs = DummyVectorEnv(
            [lambda: gym.make(args.task) for _ in range(args.test_num)]
        )
        # seed
        np.random.seed(args.seed)
        torch.manual_seed(args.seed)
        train_envs.seed(args.seed)
        test_envs.seed(args.seed)
        # Q_param = V_param = {"hidden_sizes": [128]}
        # model
        net = Net(
            args.state_shape,
            hidden_sizes=args.hidden_sizes,
            device=args.device,
            # dueling=(Q_param, V_param),
        ).to(args.device)
        
        # NOTE: just add this line
        net = Actor(net, action_shape=args.action_shape, device=args.device).to(args.device)
    
        optim = torch.optim.Adam(net.parameters(), lr=args.lr)
        policy = DQNPolicy(
            net,
            optim,
            args.gamma,
            args.n_step,
            target_update_freq=args.target_update_freq,
        )
        # buffer
        if args.prioritized_replay:
            buf = PrioritizedVectorReplayBuffer(
                args.buffer_size,
                buffer_num=len(train_envs),
                alpha=args.alpha,
                beta=args.beta,
            )
        else:
            buf = VectorReplayBuffer(args.buffer_size, buffer_num=len(train_envs))
        # collector
        train_collector = Collector(policy, train_envs, buf, exploration_noise=True)
        test_collector = Collector(policy, test_envs, exploration_noise=True)
        # policy.set_eps(1)
        train_collector.collect(n_step=args.batch_size * args.training_num)
        # log
        log_path = os.path.join(args.logdir, args.task, 'dqn')
        writer = SummaryWriter(log_path)
        logger = TensorboardLogger(writer)
    
        def save_best_fn(policy):
            torch.save(policy.state_dict(), os.path.join(log_path, 'policy.pth'))
    
        def stop_fn(mean_rewards):
            return mean_rewards >= args.reward_threshold
    
        def train_fn(epoch, env_step):
            # eps annnealing, just a demo
            if env_step <= 10000:
                policy.set_eps(args.eps_train)
            elif env_step <= 50000:
                eps = args.eps_train - (env_step - 10000) / \
                    40000 * (0.9 * args.eps_train)
                policy.set_eps(eps)
            else:
                policy.set_eps(0.1 * args.eps_train)
    
        def test_fn(epoch, env_step):
            policy.set_eps(args.eps_test)
    
        # trainer
        result = offpolicy_trainer(
            policy,
            train_collector,
            test_collector,
            args.epoch,
            args.step_per_epoch,
            args.step_per_collect,
            args.test_num,
            args.batch_size,
            update_per_step=args.update_per_step,
            train_fn=train_fn,
            test_fn=test_fn,
            stop_fn=stop_fn,
            save_best_fn=save_best_fn,
            logger=logger,
        )
        assert stop_fn(result['best_reward'])
    
        if __name__ == '__main__':
            pprint.pprint(result)
            # Let's watch its performance!
            env = gym.make(args.task)
            policy.eval()
            policy.set_eps(args.eps_test)
            collector = Collector(policy, env)
            result = collector.collect(n_episode=1, render=args.render)
            rews, lens = result["rews"], result["lens"]
            print(f"Final reward: {rews.mean()}, length: {lens.mean()}")
    
    
    def test_pdqn(args=get_args()):
        args.prioritized_replay = True
        args.gamma = .95
        args.seed = 1
        test_dqn(args)
    
    
    if __name__ == '__main__':
        test_dqn(get_args())

 ```

**Comment by link-zju** [comment]
The version of tianshou is 0.4.8.

**Comment by Trinkle23897** [comment]
I think the reason is because of the final layer's softmax operation: https://github.com/thu-ml/tianshou/blob/278c91a2228a46049a29c8fa662a467121680b10/tianshou/utils/net/discrete.py#L43

**Comment by link-zju** [comment]
Thanks a lot. It works normally with softmax_output=False. 
But why does this issue appear? The softmax seems not to influence action selection (argmax).

**Comment by Trinkle23897** [comment]
But it would make the network harder to learn. The range of output becomes [0, 1] instead of ( $-\infty$, $+\infty$) and the gradient update would vanish, I guess, at least for DQN.

**Comment by link-zju** [comment]
Thank you so much. This should be the reason. 
It reminds me that the loss value hardly changes when softmax_output=True (the default setting).

**Comment by link-zju** [comment]
By the way, I want to confirm that the softmax_output should always be **False** for DQN not only due to the scale of the gradient. 
Since the actor is an estimator of the action value function, the softmax of the logit might be meaningless.

**Comment by Trinkle23897** [comment]
Yes this is correct.

**Comment by link-zju** [comment]
All is clear. Kudos to your timely reply. 👍
